crates/ruma-server-util/src/authorization.rs

```diff
@@ -0,0 +1,198 @@
+//! Common types for implementing federation authorization
+
+use headers::{authorization::Credentials, HeaderValue};
+use ruma_common::{ServerName, ServerSigningKeyId};
+use tracing::debug;
+
+/// Typed representation of an `Authorization` header of scheme `X-Matrix`, as defined in the
+/// [Matrix Server-Server API][spec]. Includes an implementation of
+/// [`headers::authorization::Credentials`] for automatically handling the encoding and decoding
+/// when using a web framework that supports typed headers.
+///
+/// [spec]: https://spec.matrix.org/v1.3/server-server-api/#request-authentication
+#[non_exhaustive]
+pub struct XMatrix {
+    /// the server name of the sending server
+    pub origin: Box<ServerName>,
+    /// the server name of the receiving sender. For compatibility with older servers, recipients
+    /// should accept requests without this parameter, but MUST always send it. If this property is
+    /// included, but the value does not match the receiving server’s name, the receiving server
+    /// must deny the request with an HTTP status code 401 Unauthorized.
+    pub destination: Option<Box<ServerName>>,
+    ///the ID, including the algorithm name, of the sending server’s key used to sign the request
```

Comment:
```suggestion
    /// The ID - including algorithm - of the sending server's key that was used to sign the request
```

---

same @ other doc bits, capitalize the first word of a sentence, and please convert all the smart quotes in the docs to `'` and `"` :pray:

Reply:
This was taken straight from the matrix spec, but yeah, can still do this.